### PR TITLE
HTCONDOR-2274  apptainer home dir

### PR DESCRIPTION
**Pull request by GregThain**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: docs/version-history/feature-versions-23-x.rst [hunk 1]
```diff
@@ -24,6 +24,13 @@ New Features:
   The given file is overwritten instead of appended to.
   :jira:`2425`
 
+- Container universe jobs running under singularity or apptainer now
+  run with a contained home directory, when HTCondor file transfer is
+  enabled.  This means the jobs get the $HOME environment variable set
+  to the scratch directory, and /etc/passwd entry inside the container
```
Comment (Todd-L-Miller) [inline docs/version-history/feature-versions-23-x.rst:30]:
```suggestion
  to the scratch directory, and an /etc/passwd entry inside the container
```
?

**Review by Todd-L-Miller — COMMENTED** [review]
The code looks fine, but what's the reasoning behind only doing it when `job_iwd` is the same as `execute_dir`, and why does the release note claim that happens if and only if file transfer is enabled?  (I guess because we don't create a sandbox directory in that case?  Is the logic here that jobs shouldn't have access to a user's shared FS home directory?)

**Comment by GregThain** [comment]
> The code looks fine, but what's the reasoning behind only doing it when `job_iwd` is the same as `execute_dir`, and why does the release note claim that happens if and only if file transfer is enabled?

In this place in the starter, we don't have a flag telling us if we've done file transfer.  The starter always makes a scratch directory, even when file transfer is on.  For example, we still drop the .job.ad and .machine.ad into the scratch directory when file transfer is off.  When file transfer is on, the cwd of the job starts out as the scratch directory created by the starter.  When file transfer is off, the cwd of the job is usually the pwd of the AP where condor_submit was run.  When file xfer is off, the cwd of the job can never be the scratch directory.

So, the canonical way in the starter to detect if file xfer is on is to compare `job_iwd` (the pwd the job starts in) with `execute_dir` (the scratch dir).  If they are the same, we assume file xfer is on.

In this case, the assumption is that when file xfer is on, the job runs with some arbitrary uid unknown at container build time, so the job can't read/write from/to some home directory known a-priori, so we want to map any home directory access into scratch.

If file transfer is off, the job is probably running with the submitting uid, and might have the EP's (and thus also the AP's) home directory bind-mounted into the container, so we don't do that.

**Comment by Todd-L-Miller** [comment]
> In this place in the starter, we don't have a flag telling us if we've done file transfer.
[snip]
> So, the canonical way in the starter to detect if file xfer is on is to compare `job_iwd` (the pwd the job starts in) with `execute_dir` (the scratch dir). If they are the same, we assume file xfer is on.

OK, that's sad but fine.
 
> In this case, the assumption is that when file xfer is on, the job runs with some arbitrary uid unknown at container build time, so the job can't read/write from/to some home directory known a-priori, so we want to map any home directory access into scratch.

OK, this makes sense.  I don't see any immediate need for a knob.

**Review by Todd-L-Miller — APPROVED** [review]

**Comment by Todd-L-Miller** [comment]
* JIRA ticket still needs a target version.
* I assume that it will be 23.x or more specific, in which case the branch will match.
* Since we ship a no-root-required apptainer with HTCondor, this seems eminently testable.

**Pull request merged**